lib/ast-utils.js

```diff
@@ -0,0 +1,186 @@
+                return findJSDocComment(node.leadingComments, line);
+            }
+            break;
+
+        case "ArrowFunctionExpression":
+        case "FunctionExpression":
+
+            if (parent.type !== "CallExpression") {
+                while (parent && !parent.leadingComments && !/Function/.test(parent.type)) {
+                    parent = parent.parent;
+                }
+
+                return parent && (parent.type !== "FunctionDeclaration") ? findJSDocComment(parent.leadingComments, line) : null;
+            }
+
+        // falls through
+
+        default:
+            return null;
+    }
+}
+
+/**
+ * Determines whether two adjacent tokens are have whitespace between them.
+ * @param {Object} left - The left token object.
+ * @param {Object} right - The right token object.
+ * @returns {boolean} Whether or not there is space between the tokens.
+ * @public
+ */
+function isTokenSpaced(left, right) {
+    return left.range[1] < right.range[0];
+}
+
+/**
+ * Determines whether two adjacent tokens are on the same line.
+ * @param {Object} left - The left token object.
+ * @param {Object} right - The right token object.
+ * @returns {boolean} Whether or not the tokens are on the same line.
+ * @public
+ */
+function isTokenOnSameLine(left, right) {
+    return left.loc.end.line === right.loc.start.line;
+}
+
+/**
+ * Checks whether or not a node is `null` or `undefined`.
+ * @param {ASTNode} node - A node to check.
+ * @returns {boolean} Whether or not the node is a `null` or `undefined`.
+ * @public
+ */
+function isNullOrUndefined(node) {
+    return (
+        (node.type === "Literal" && node.value === null) ||
+        (node.type === "Identifier" && node.name === "undefined") ||
+        (node.type === "UnaryExpression" && node.operator === "void")
+    );
+}
+
+/**
+ * Checks reference if is non initializer and writable.
+ * @param {Reference} reference - A reference to check.
+ * @param {int} index - The index of the reference in the references.
+ * @param {Reference[]} references - The array that the reference belongs to.
+ * @returns {boolean} Success/Failure
+ * @private
+ */
+function isModifyingReference(reference, index, references) {
+    var identifier = reference.identifier;
+
+    return (identifier != null &&
+        reference.init === false &&
+        reference.isWrite() &&
+            // Destructuring assignments can have multiple default value,
+            // so possibly there are multiple writeable references for the same identifier.
+        (index === 0 || references[index - 1].identifier !== identifier)
+    );
+}
+
+/**
+ * Gets references which are non initializer and writable.
+ * @param {Reference[]} references - An array of references.
+ * @returns {Reference[]} An array of only references which are non initializer and writable.
+ * @public
+ */
+function getModifyingReferences(references) {
+    return references.filter(isModifyingReference);
+}
+
+
+module.exports = {
```

Comment:
Still would like it so the functions are directly on this object instead of references, its too easy to forget to export functions this way.